src/FetchInterceptor.js

```diff
@@ -0,0 +1,327 @@
+    const { shouldIntercept } = this.config;
+
+    return Promise.resolve(shouldIntercept(request))
+      .then(shouldIntercept =>
+        ({ ...requestUnit, shouldIntercept })
+      );
+  }
+
+  authorizeRequest(requestUnit) {
+    const { shouldIntercept } = requestUnit;
+
+    if (!shouldIntercept) {
+      return requestUnit;
+    }
+
+    const { request } = requestUnit;
+    const { accessToken } = this.accessTokenProvider.getAuthorization();
+    const { authorizeRequest } = this.config;
+
+    if (request && accessToken){
+      return Promise.resolve(authorizeRequest(request, accessToken))
+        .then(request =>
+          ({ ...requestUnit, accessToken, request })
+        );
+    }
+
+    return requestUnit;
+  }
+
+  shouldFetch(requestUnit) {
+    const { request } = requestUnit;
+    const { shouldFetch } = this.config;
+
+    // verifies all outside conditions from config are met
+    if (!shouldFetch) {
+      return requestUnit;
+    }
+
+    return Promise.resolve(shouldFetch(request))
+      .then(shouldFetch =>
+        ({ ...requestUnit, shouldFetch })
+      );
+  }
+
+  fetchRequest(requestUnit) {
+    const { shouldFetch } = requestUnit;
+
+    if (shouldFetch) {
+      const { request, fetchCount } = requestUnit;
+      const { fetchRetryCount } = this.config;
+
+      // verifies that retry count has not been exceeded
+      if (fetchCount > fetchRetryCount) {
+        throw new RetryCountExceededException(requestUnit);
+      }
+
+      const { fetch } = this;
+      return Promise.resolve(fetch(request))
+        .then(response =>
+          ({
+            ...requestUnit,
+            response,
+            fetchCount: fetchCount + 1,
+          })
+        );
+    }
+
+    return requestUnit;
+  }
+
+  shouldInvalidateAccessToken(requestUnit) {
+    const { shouldIntercept } = requestUnit;
+    const { shouldInvalidateAccessToken } = this.config;
+
+    if (shouldIntercept && shouldInvalidateAccessToken) {
+      const { response } = requestUnit;
+      // check if response invalidates access token
+      return Promise.resolve(shouldInvalidateAccessToken(response))
+        .then(shouldInvalidateAccessToken =>
+          ({ ...requestUnit, shouldInvalidateAccessToken })
+        );
+    }
+
+    return requestUnit;
+  }
+
+  invalidateAccessToken(requestUnit) {
+    const { shouldIntercept, shouldInvalidateAccessToken } = requestUnit;
+
+    if (shouldIntercept && shouldInvalidateAccessToken) {
+      this.accessTokenProvider.renew();
+    }
+
+    return requestUnit;
+  }
+
+  handleResponse(requestUnit) {
+    const { shouldIntercept, response, fetchResolve, fetchReject } = requestUnit;
+
+    // can only be empty on network errors
+    if (!response) {
+      fetchReject();
+      return;
+    }
+
+    if (shouldIntercept && isResponseUnauthorized(response)) {
+      throw new TokenExpiredException({ ...requestUnit })
+    }
+
+    if (this.config.onResponse) {
+      this.config.onResponse(response);
```

Comment:
What if someone in onResponse function doesn't know how to handle `response` with `clone` is it going to destroy it before `fetchResolve`, should you pass response.clone here?